### ISSUE TITLE
Consensus Committee by-view / by-epoch queries should support all known epochs

### ISSUE THREAD

**Issue by jordanschalm** [onflow/flow-go#4085]
## Context

The `Epochs` API in the protocol state only supports querying either the `Current`, `Next`, or `Previous` epoch, with respect to a particular reference block. This is based on the assumption that any node only needs to know about these 3 epochs. The Consensus Committee and EpochLookup modules follow this pattern.

However, the use of these modules has expanded. For example, an access node might look up the consensus committee by view for an epoch several weeks ago, to satisfy by block request by ID or height. For this use-case, it is necessary to:
- effectively query epochs by view, independent of a reference block
- query epoch information for epochs relatively far in the past

@durkmurder has implemented a hot-fix for this issue in https://github.com/onflow/flow-go/pull/4063.

## Definition of Done
- Update `ByView` methods in consensus Committee to return a value for all views within an epoch processed by the node.
- Add test case for historical epoch query to Committee tests

**Comment by AlexHentschel** [comment]
I am curious to hear what implementation approach you had in mind. I think we should pay attention to the following details:
* at the moment, `committees.Consensus` operates purely in memory and event driven:
  
   It consumes `EpochCommittedPhaseStarted` notifications and updates its internal memory that only holds 3 Epochs.
     As it only needs to inspect 3 in-memory structs, it has a **bounded runtime** without needing any potentially blocking IO.

A bounded runtime with a nearly instant return is a very important property of the implementation that we have to retain for security reasons: 
* If the implementation falls back on a database call or maintains a large or possibly unbounded number of Epochs in memory, byznatine nodes can potentially mount a resource exhaustion attack on lines: https://github.com/onflow/flow-go/blob/7fdc7fbdbd353d4d1f46c8240a4c310c81b730f9/consensus/hotstuff/timeoutaggregator/timeout_aggregator.go#L156-L162 (and probably other parts of the consensus code too).

  Potential Attack Scenario:
    * Evil Edward floods Alice with made-up Timeout Objects [`TO`s] for various views that are very far in the future
    * In [line 156](https://github.com/onflow/flow-go/blob/7fdc7fbdbd353d4d1f46c8240a4c310c81b730f9/consensus/hotstuff/timeoutaggregator/timeout_aggregator.go#L156) Alice tries to instantiate a `TimeoutCollector` for the future view, which through some layers of abstraction will call the factory method for `TimeoutProcessor` https://github.com/onflow/flow-go/blob/7fdc7fbdbd353d4d1f46c8240a4c310c81b730f9/consensus/hotstuff/timeoutcollector/factory.go#L82-L83
    * It tries to retrieve the Identities for the `TO`'s from the `committees.Consensus`
    * Lets assume that [line 83](https://github.com/onflow/flow-go/blob/7fdc7fbdbd353d4d1f46c8240a4c310c81b730f9/consensus/hotstuff/timeoutcollector/factory.go#L83) falls back to a data base call before the implementation realizes that the view is way too far in the future and returns a `model.ErrViewForUnknownEpoch` 
       
       Due to the latency of a data base read, an attacker can very effectively keep all worker threads of the `TimeoutAggregator` busy with data base queries, and thereby prevent Alice from effectively participating in TC generation. 


The important take away is: 
* `committees.Consensus` is a critical component for deciding whether some message is valid or not. Therefore, it is part of the first line of defence against resource exhaustion attacks. The resulting requirement is that  `committees.Consensus` must handle an abnormally high number of queries with minimal system load. We want to be confident that the implementation can handle any malicious load up to the network connection becoming the ultimate bottleneck.
* More generally: if you are thinking about adding a data base read somewhere on the hot-path of consensus, you should pause and ask the following question: "What is the argument that this data base call cannot be utilized for a resource exhaustion attack?" You need a conclusive answer to this question before you can proceed with the implementation. 


Overall, I feel there are two largely disjoint use cases for `committees.Consensus`:
1. Serving the hot-path of consensus for confirming whether or not incoming messages are valid / processable. This needs to be fully in-memory with bounded runtime and nearly immediate returns to mitigate resource exhaustion attacks. 
2.  serving the access API, as part of which db queries are common practise. Protection from resource exhaustion attacks is guaranteed by global rate limits or trust relationships. 

Therefore, I feel completely independent implementations are warranted.

**Comment by AlexHentschel** [comment]
I find the estimate of 3SP a bit optimistic. Would tend more towards 5 including review and discussion

**Comment by jordanschalm** [comment]
I agree with your points about avoiding introducing attack vectors in the main HotStuff message validation code path. The main problem this issue was originally trying to address is summarized in this TODO:

https://github.com/onflow/flow-go/blob/8151d1963b9dedbf8f48e701373a3828c26b8eb8/consensus/hotstuff/signature/block_signer_decoder.go#L43-L46

The concern was, with the Dynamic Protocol State, the query we were falling back to could return a different result than the happy-path query, since it incorporates block-by-block identity table changes, which would cause incorrect behaviour. 

Specifically, if a node is ejected, it will appear in the `ByView` query but not in the `ByBlockID` query. 

#### Categorization
I would categorize this as a bug rather than as tech debt, as it will cause incorrect results to be returned by the Access API, if a Consensus node is ejected. We can deprioritize it until Ejection is being implemented, but not later than that.

#### Implementation Approach

> I am curious to hear what implementation approach you had in mind

What I had in mind initially was something like:
- The Protocol State maintains an index of view to epoch counter (probably a key prefix containing view numbers that can be traversed to implement the lookup)
  - The index is populated when an `EpochSetup` event is sealed
- The Consensus Committee remains largely the same, but falls back to using the index for `ByView` queries, for known epochs, rather than returning an error.

#### A simpler solution for the underlying problem
Another option to address the problem with the `BlockSignerDecoder`, which would be much simpler, is to _continue_ falling back to `state.AtBlockID(...).Identities` but with a different filter than what is used by `committees.Consensus`. 

I have the impression that a general-purpose `view->epoch` lookup that isn't limited to the 3 most recent ones would be useful beyond this issue, but if it isn't, then the suggestion above would be sufficient. 

> at the moment, committees.Consensus operates purely in memory and event driven:

The `ByView` queries are fully in-memory and event-driven, the `ByBlockID` queries are not: they [fallback to the protocol state](https://github.com/onflow/flow-go/blob/b2c97b5696506c80236821e1856c453e947f634f/consensus/hotstuff/committees/consensus_committee.go#L197) which will query the DB for uncached requests.
